### PR TITLE
Add type annotations for CLIP (torch) (#16059)

### PR DESCRIPTION
**Pull request by jacobdineen**
# What does this PR do?

Added type annotations to CLIP per the guidelines. Ran make fixup and ran tests before creating PR.

#16059 

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. [link here](https://github.com/huggingface/transformers/issues/16059)

### CODE REVIEW

**Comment by debtriche** [comment]
No emails

I’m about to report you people for harassment 

Stop fucking emailing me !!!!!!!!!!

Fucking Assholes

**Comment by Rocketknight1** [comment]
@debtriche I'm afraid that's not our fault - I'm guessing you've accidentally "watched" the Transformers repo! If you unselect that, you should stop receiving that stream of e-mails. 
image

File: src/transformers/models/clip/modeling_clip.py [hunk 1]
```diff
@@ -95,12 +95,12 @@ class CLIPOutput(ModelOutput):
...
-    text_embeds: torch.FloatTensor = None
-    image_embeds: torch.FloatTensor = None
-    text_model_output: BaseModelOutputWithPooling = None
-    vision_model_output: BaseModelOutputWithPooling = None
+    logits_per_image: Optional[torch.FloatTensor] = None
+    logits_per_text: Optional[torch.FloatTensor] = None
+    text_embeds: Optional[torch.FloatTensor] = None
+    image_embeds: Optional[torch.FloatTensor] = None
+    text_model_output: Optional[BaseModelOutputWithPooling] = None
+    vision_model_output: Optional[BaseModelOutputWithPooling] = None
```
Comment (Rocketknight1) [inline src/transformers/models/clip/modeling_clip.py:103]: I don't think the `Optional` type is necessary for dataclass definitions, just function headers!

Comment (jacobdineen) [reply src/transformers/models/clip/modeling_clip.py]: revised

**Review by Rocketknight1 — COMMENTED** [review]
Thanks for this! I added some suggestions, and we should be good to go once they're addressed!

File: src/transformers/models/clip/modeling_clip.py [hunk 2]
```diff
@@ -953,15 +958,15 @@ def get_image_features(
...
-    ):
+        input_ids: Optional[torch.LongTensor] = None,
+        pixel_values: Optional[torch.FloatTensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        return_loss: Optional[bool] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+    ) -> Union[Tuple, CLIPOutput]:
```
Comment (jacobdineen) [inline src/transformers/models/clip/modeling_clip.py:969]:
```suggestion
    ) -> Union[Tuple, CLIPOutput]:
```

Fixes committed. I am unsure on this one, as I assume that this returns a different output type than ClipOutput instance

Comment (jacobdineen) [reply src/transformers/models/clip/modeling_clip.py:969]: awesome - thanks for your help! Little bit of a learning curve..

Tests passed, and make fixup called.

**Review by Rocketknight1 — APPROVED** [review]
It looks perfect! Thanks for the PR, and the quick turnaround on the comments!

**Comment by debtriche** [comment]
Thank:)

I’ve been trying everything, it’s absolutely been driving me mad

**Pull request merged**